### PR TITLE
Better handle 404 response for federation /send/

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/util/retryutils.py [hunk 1]
```diff
@@ -116,14 +117,24 @@ def __init__(self, destination, clock, store, retry_interval,
...
+        self.backoff_on_404 = backoff_on_404
 
     def __enter__(self):
         pass
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         valid_err_code = False
         if exc_type is not None and issubclass(exc_type, CodeMessageException):
-            valid_err_code = exc_val.code != 429 and 0 <= exc_val.code < 500
+            if exc_val.code < 400:
```
Comment (NegativeMjark) [inline synapse/util/retryutils.py]: Maybe chuck in a short comment explaiing what's going on here?

File: synapse/util/retryutils.py [hunk 2]
```diff
@@ -116,14 +117,30 @@ def __init__(self, destination, clock, store, retry_interval,
...
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         valid_err_code = False
         if exc_type is not None and issubclass(exc_type, CodeMessageException):
-            valid_err_code = exc_val.code != 429 and 0 <= exc_val.code < 500
+            # Some error codes are perfectly fine for some APIs, whereas other
+            # APIs may expect to never received e.g. a 404. It's important to
+            # handle 404 as some remote servers will return a 404 when the HS
+            # has been decommissioned.
+            if exc_val.code < 400:
```
Comment (NegativeMjark) [inline synapse/util/retryutils.py]: Isn't the `exc_val.code < 400` redundant because of the `exc_val.code < 500` ?

**Comment by NegativeMjark** [comment]
LGTM apart from the possibly redundant check?

File: synapse/util/retryutils.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/util/retryutils.py]: Yeah, though personally I quite like the fact that each case of code is handled in order. I can change it if you think it makes it clearer?

Comment (NegativeMjark) [inline synapse/util/retryutils.py]: I find it misleading because it makes it seem as though 4xx are handled differently to 2xx/3xx.

**Comment by erikjohnston** [comment]
@matrixbot retest this please

**Pull request merged**